components/tools/OmeroPy/src/omero/plugins/metadata.py

```diff
@@ -554,6 +565,87 @@ def populateroi(self, args):
                 meas = ctx.get_measurement_ctx(i)
                 meas.parse_and_populate()
 
+    def pixelsize(self, args):
+        "Set physical pixel size"
+        md = self._load(args)
+        client, conn = self._clientconn(args)
+
+        units = {
+            'nm': UnitsLength.NANOMETER,
+            'um': UnitsLength.MICROMETER,
+            'mm': UnitsLength.MILLIMETER,
+            'cm': UnitsLength.CENTIMETER,
+            'm': UnitsLength.METER,
+            'km': UnitsLength.KILOMETER
+        }
+
+        if args.unit not in units:
+            self.ctx.die(100, "%s not recognized as valid unit." % args.unit)
+
+        if not args.x and not args.y and not args.z:
+            self.ctx.die(100, "No pixel sizes specified.")
+
+        if md.get_type() == "Screen":
+            q = """SELECT pix FROM Pixels pix, WellSample ws, Plate p,
+                   ScreenPlateLink spl WHERE
+                   spl.child=p AND pix.image=ws.image AND ws.well.plate=p AND
+                   spl.parent.id=:id"""
+        elif md.get_type() == "Plate":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.well.plate.id=:id"""
+        elif md.get_type() == "PlateAcquisition":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.plateAcquisition.id=:id"""
+        elif md.get_type() == "Well":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.well.id=:id"""
+        elif md.get_type() == "Project":
+            q = """SELECT pix FROM Pixels pix, DatasetImageLink dil,
+                   ProjectDatasetLink pdl WHERE dil.child=pix.image AND
+                   dil.parent=pdl.child AND pdl.parent.id=:id"""
+        elif md.get_type() == "Dataset":
+            q = """SELECT pix FROM Pixels pix, DatasetImageLink dil WHERE
+                   dil.child=pix.image AND dil.parent.id=:id"""
+        elif md.get_type() == "Image":
+            q = """SELECT pix FROM Pixels pix WHERE pix.image.id=:id"""
+        else:
+            raise Exception("Not implemented for type %s" % md.get_type())
+
+        params = omero.sys.ParametersI()
+        params.addId(md.get_id())
+        pixels = client.getSession().getQueryService()\
+                                    .findAllByQuery(q, params)
+        for pixel in pixels:
+            if args.x:
+                pixSize = pixel.getPhysicalSizeX()
+                if not pixSize:
```

Comment:
to avoid copy/paste of similar code, you should introduce a method to the set the physical size